crates/bevy_ecs/src/component.rs

```diff
@@ -228,7 +228,7 @@ use derive_more::derive::{Display, Error};
 ///
 /// In general, this shouldn't happen often, but when it does the algorithm is simple and predictable:
 /// 1. Use all of the constructors (including default constructors) directly defined in the spawned component's require list
-/// 2. In the order the requires are defined in `#[require()]`, recursively visit the require list of each of the components in the list (this is a Depth First Search). When a constructor is found, it will only be used if one has not already been found.
+/// 2. In the order the requires are defined in `#[require()]`, recursively visit the require list of each of the components in the list (this is a Breadth First Search). When a constructor is found, it will only be used if one has not already been found.
```

Comment:
That sounds a bit off. Maybe:

> ...a direct `#[require()]` are...

Reply:
That sounds a bit off. Maybe:

> /// The algorithm for choosing the constructor from the tree:
> /// 1. A constructor from a direct #[requires()], if one exists, is selected with priority.
> /// 2. Otherwise, perform a Depth First Search on the tree of requirements and select the first one found.

Reply:
I think the main issue with the current wording is that step 1 refers to the "spawned component", which seems to suggest that the step only applies to the component you directly spawn (aka the root of the tree), when it actually applies to every component visited during the DFS. I'd reword it to something roughly like this:

> 1. Starting with the spawned component, recusively visit the require list of each of the components in the list, in the order the requires are defined in `#[require()]` (this is a Depth First Search).
> 2. Whenever a component is visited, use all of the constructors (including default constructors) directly defined in its require list, except for the components that have already been found.

Reply:
On my readthrough of it I thought "depth-first" was somewhat misleading, but maybe I'm just misunderstanding the term.

Reply:
The implementation is actually a depth-first search, it just collects the constructors whenever it visits a component.